src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -1405,26 +1407,111 @@ private Type EmitFunctionCallNode(AssociativeNode node)
+                EmitOpCode(OpCodes.Call, keygen);
+
+                var local = DeclareLocal(typeof(IEnumerable<ProtoCore.CLRFunctionEndPoint>), "cached MethodBase objects");
+                //local could be null if we are not emitting currently.
+                if (local != null)
                 {
-                    Type t = DfsTraverse(arg);
+                    EmitOpCode(OpCodes.Ldloca, local);
+                }
+
+                // Emit methodCache.TryGetValue(KeyGen(...), out IEnumerable<CLRFunctionEndPoint> feps)
+                var dictLookup = typeof(IDictionary<int, IEnumerable<ProtoCore.CLRFunctionEndPoint>>).GetMethod(
+                    nameof(IDictionary<int, IEnumerable<ProtoCore.CLRFunctionEndPoint>>.TryGetValue));
+                EmitOpCode(OpCodes.Callvirt, dictLookup);
+
+                EmitOpCode(OpCodes.Pop);
+                if (local != null)
+                {
+                    EmitOpCode(OpCodes.Ldloc, local.LocalIndex);
+                }
+
+                //emit an array of args to pass to unmarshal
+                EmitArray(typeof(object), args, (AssociativeNode n, int index) =>
+                {
+                    Type t = DfsTraverse(n);
                     if (isStaticOrCtor)
                     {
-                        t = EmitCoercionCode(arg, t, parameters[index]);
+                        t = EmitCoercionCode(n, t, parameters[index]);
                     }
                     else if (index > 0)
                     {
-                        t = EmitCoercionCode(arg, t, parameters[index - 1]);
+                        t = EmitCoercionCode(n, t, parameters[index - 1]);
+                    }
+
+                    if (t == null) return;
+
+                    if (t.IsValueType)
+                    {
+                        EmitOpCode(OpCodes.Box, t);
+                    }
+                });
+
+                //TODO if possible only marshal args that resulted from a replicated call.
+                
+                //emit marshal for args
+
+                // Emit call to load the runtimeCore argument
+                EmitOpCode(OpCodes.Ldarg_3);
+                //TODO cache this at start.
+                //now call marshall to convert any clrstackvalues to plain c# objs. 
+                var marhshallMethod = typeof(Replication).GetMethod(nameof(Replication.UnMarshalFunctionArguments2), BindingFlags.Static | BindingFlags.Public);
+                EmitOpCode(OpCodes.Call, marhshallMethod);
+
+                var unMarshaledArgsArray = DeclareLocal(marhshallMethod.ReturnType, "unMarshaledArgsArray");
+                if (unMarshaledArgsArray != null)
+                {
+                    EmitOpCode(OpCodes.Stloc, unMarshaledArgsArray.LocalIndex);
+                }
+                //now on the stack we have a list of objects, we need to iterate through it and push each item to the stack.
+                //emit a for loop which pushes to stack.
```

Comment:
I believe the stloc in the previous line will pop the array from the stack.

Reply:
thats @aparajit-pratap that makes sense - it should be okay though since we do 
`EmitOpCode(OpCodes.Ldloc, unMarshaledArgsArray.LocalIndex);` at each step in the unrolling below.

Though I should definitely clarify the comment.